In [13]:
%pip install nltk 
import pandas as pd 
from nltk import tokenize  
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.probability import FreqDist 
nltk.download('stopwords')
import string


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.0.1 -> 23.3
[notice] To update, run: python.exe -m pip install --upgrade pip
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\fobus\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [38]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize


nltk.download('punkt')
nltk.download('stopwords') 
nltk.download('averaged_perceptron_tagger')



[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\fobus\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\fobus\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\fobus\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping taggers\averaged_perceptron_tagger.zip.


True

In [57]:
Myarray= ['C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/The Hobbit 1937.txt','C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/Two Towers 1954.txt', 
    'C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/The Fellowship of the Ring 1954.txt','C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/THE ADVENTURES OF TOM BOMBADIL 1967.txt', 
    'C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/The Silmarillion 1977.txt','C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/The Children 2007.txt'] 
 
import nltk
from nltk.tokenize import word_tokenize
from nltk.tag import pos_tag

# with open(i, 'r', encoding='utf-8') as file:
#     text = file.read()
#     text = text.lower()
#     tokens = word_tokenize(text)

# tagged_words = pos_tag(tokens)

# pos_counts = {'VBZ': 0, 'VBN': 0, 'VB': 0, 'VBD': 0, 'VBG': 0, 'VBP': 0, 
#               'NN': 0, 'NNS': 0, 'NNP': 0, 
#               'JJ': 0, 'JJR': 0, 'JJS': 0}

# for word, tag in tagged_words:
#     if tag in pos_counts:
#         pos_counts[tag] += 1

# print("Verb Counts:")
# for pos in ['VBZ', 'VBN', 'VB', 'VBD', 'VBG', 'VBP']:
#     print(f'{pos}: {pos_counts[pos]}')

# print("\nNoun Counts:")
# for pos in ['NN', 'NNS', 'NNP']:
#     print(f'{pos}: {pos_counts[pos]}')

# print("\nAdjective Counts:")
# for pos in ['JJ', 'JJR', 'JJS']:
#     print(f'{pos}: {pos_counts[pos]}')
    
    

In [59]:
for i in  Myarray :  
    with open(i, 'r', encoding='utf-8') as file:
        text = file.read()
        text = text.lower()
        tokens = word_tokenize(text)

    tagged_words = pos_tag(tokens)

    pos_counts = {'VBZ': 0, 'VBN': 0, 'VB': 0, 'VBD': 0, 'VBG': 0, 'VBP': 0, 
              'NN': 0, 'NNS': 0, 'NNP': 0, 
              'JJ': 0, 'JJR': 0, 'JJS': 0}

    for word, tag in tagged_words:
        if tag in pos_counts:
            pos_counts[tag] += 1 
 
 
    print(i)
    print("Глаголы:")
    for pos in ['VBZ', 'VBN', 'VB', 'VBD', 'VBG', 'VBP']:
        print(f'{pos}: {pos_counts[pos]}')

    print("\nСуществительные:")
    for pos in ['NN', 'NNS', 'NNP']:
        print(f'{pos}: {pos_counts[pos]}')

    print("\nПрилагательные:")
    for pos in ['JJ', 'JJR', 'JJS']:
        print(f'{pos}: {pos_counts[pos]}') 
         
    total_words = sum(pos_counts.values())
    print(f"\nОбщее количество слов: {total_words}")


    noun_percentage = (pos_counts.get('NN', 0) + pos_counts.get('NNS', 0) + pos_counts.get('NNP', 0)) / total_words * 100
    adj_percentage = (pos_counts.get('JJ', 0) + pos_counts.get('JJR', 0) + pos_counts.get('JJS', 0)) / total_words * 100
    verb_percentage = (pos_counts.get('VBZ', 0) + pos_counts.get('VBN', 0) + pos_counts.get('VB', 0) + pos_counts.get('VBD', 0) + pos_counts.get('VBG', 0) + pos_counts.get('VBP', 0)) / total_words * 100

    print(f"\nПроцентное соотношение существительных: {noun_percentage:.2f}%")
    print(f"Процентное соотношение прилагательных: {adj_percentage:.2f}%")
    print(f"Процентное соотношение глаголов: {verb_percentage:.2f}%")
    

C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/The Hobbit 1937.txt
Глаголы:
VBZ: 985
VBN: 2392
VB: 3806
VBD: 7710
VBG: 1927
VBP: 1655

Существительные:
NN: 14190
NNS: 5210
NNP: 12

Прилагательные:
JJ: 6480
JJR: 305
JJS: 229

Общее количество слов: 44901

Процентное соотношение существительных: 43.23%
Процентное соотношение прилагательных: 15.62%
Процентное соотношение глаголов: 41.15%
C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/Two Towers 1954.txt
Глаголы:
VBZ: 2551
VBN: 3747
VB: 7162
VBD: 11316
VBG: 2801
VBP: 3512

Существительные:
NN: 24638
NNS: 7639
NNP: 53

Прилагательные:
JJ: 10673
JJR: 550
JJS: 260

Общее количество слов: 74902

Процентное соотношение существительных: 43.16%
Процентное соотношение прилагательных: 15.33%
Процентное соотношение глаголов: 41.51%
C:/Users/fobus/Desktop/GIT_TEST/nltk/HomeWork2/The Fellowship of the Ring 1954.txt
Глаголы:
VBZ: 2793
VBN: 4805
VB: 8199
VBD: 14590
VBG: 3343
VBP: 3985

Существительные:
NN: 30603
NNS: 8725
NNP: 43

Прилагательные:
JJ: 12

Глаголы:  
VBZ (3-е лицо ед.ч.),  
VBN (прошедшее причастие),  
VB (базовая форма глагола),  
VBD (прошедшее время),  
VBG (глагол-герундий),  
VBP (неопределенная форма глагола).<br>
Существительные:  
NN (ед.ч.),  
NNS (мн.ч.),  
NNP (ед.ч. собственное имя).<br>
Прилагательные:  
JJ (ед.ч.),  
JJR (сравнительная степень), <br>
JJS (превосходная степень).

Из результатов анализа можно сделать вывод, что со временем процентное соотношение существительных в текстах Толкиена остается стабильным, колеблясь от 43.16% до 53.73%. Это может указывать на то, что использование существительных является постоянным и важным элементом его стиля на протяжении разных периодов его творчества.

Однако, процентное соотношение прилагательных и глаголов в текстах Толкиена имеет некоторые изменения со временем. Процентное соотношение прилагательных составляет от 13.22% до 16.62%, что может указывать на то, что с течением времени Толкиен стал больше использовать прилагательные для создания описательных образов и атмосферы в своих произведениях.

С другой стороны, процентное соотношение глаголов колеблется от 32.86% до 41.51%. Это может указывать на то, что в более ранних произведениях Толкиена было больше акцента на передачу действия и движения через глаголы, в то время как в более поздних произведениях этот акцент стал несколько снижаться.